sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsGdsStore.scala

```diff
@@ -39,80 +48,166 @@ class RapidsGdsStore(
+    this.synchronized {
+      if (deviceBuffer.getLength > batchWriteBufferSize - batchWriteBufferOffset) {
+        val path = batchWriteFile.getAbsolutePath
+        withResource(new CuFileWriteHandle(path)) { handle =>
+          handle.write(batchWriteBuffer, batchWriteBufferOffset, 0)
+          logDebug(s"Spilled to $path 0:$batchWriteBufferOffset via GDS")
+        }
+        pendingBuffers.foreach(_.flush())
+        pendingBuffers.clear
+        batchWriteFile = TempSpillBufferId().getDiskPath(diskBlockManager)
+        batchWriteBufferOffset = 0
+      }
+
+      val currentBufferOffset = batchWriteBufferOffset
+      batchWriteBuffer.copyFromMemoryBuffer(
+        batchWriteBufferOffset, deviceBuffer, 0, deviceBuffer.getLength, Cuda.DEFAULT_STREAM)
+      batchWriteBufferOffset += alignUp(deviceBuffer.getLength)
+
+      val id = other.id
+      addBatchedBuffer(batchWriteFile, id)
+      val gdsBuffer = new RapidsGdsBuffer(id, batchWriteFile, currentBufferOffset, other.size,
+        other.meta, other.getSpillPriority, other.spillCallback, isPending = true)
+      pendingBuffers += gdsBuffer
+      gdsBuffer
+    }
+
+  private def alignUp(length: Long): Long = {
+    (length + blockSize - 1) & ~(blockSize - 1)
+  }
+
+  private def copyToBuffer(
+      buffer: MemoryBuffer, offset: Long, size: Long, stream: Cuda.Stream): Unit = {
+    buffer.copyFromMemoryBuffer(0, batchWriteBuffer, offset, size, stream)
+  }
+
+  private def addBatchedBuffer(path: File, id: RapidsBufferId): Set[RapidsBufferId] = {
+    val updater = new BiFunction[File, Set[RapidsBufferId], Set[RapidsBufferId]] {
+      override def apply(key: File, value: Set[RapidsBufferId]): Set[RapidsBufferId] = {
+        if (value == null) {
+          Set(id)
+        } else {
+          value + id
         }
-      } else {
-        CuFile.writeDeviceBufferToFile(path, 0, deviceBuffer)
-        0
       }
-      logDebug(s"Spilled to $path $fileOffset:${other.size} via GDS")
-      new RapidsGdsBuffer(id, fileOffset, other.size, other.meta, other.getSpillPriority,
-        other.spillCallback)
     }
+    batchedBuffers.compute(path, updater)
+  }
+
+  private def removeBatchedBuffer(path: File, id: RapidsBufferId): Set[RapidsBufferId] = {
+    val updater = new BiFunction[File, Set[RapidsBufferId], Set[RapidsBufferId]] {
+      override def apply(key: File, value: Set[RapidsBufferId]): Set[RapidsBufferId] = {
+        val newValue = value - id
+        if (newValue.isEmpty) {
+          null
+        } else {
+          newValue
+        }
+      }
+    }
+    batchedBuffers.computeIfPresent(path, updater)
   }
 
   class RapidsGdsBuffer(
       id: RapidsBufferId,
+      path: File,
       val fileOffset: Long,
       size: Long,
       meta: TableMeta,
       spillPriority: Long,
-      spillCallback: RapidsBuffer.SpillCallback)
+      spillCallback: RapidsBuffer.SpillCallback,
+      var isPending: Boolean)
```

Comment:
After thinking about the general design some more, I don't think this, or the previous `Option` approach, is going to work in light of multithreaded processing and spill. Consider the following scenario:
1. Buffer is asked to materialize/copy the buffer
1. It checks the pending flag and sees that it is pending
1. Current thread that will materialize/copy the pending buffer gets preempted
1. Another thread triggers a spill and causes the batched buffer to flush
1. Data from another buffer starts being written to the big batch buffer
1. Original thread now resumes and proceeds to copy from the big batch buffer but risks copying unrelated data, corrupting the result

There needs to be synchronized access to the pending set to determine safely whether a buffer is pending and its data is guaranteed to be in the batched buffer for the duration of a transfer.  I don't think poking a flag/option asynchronously in each buffer is going to be safe.

Reply:
Yes, that's why I added the synchronized blocks to the copying/flushing methods. The store can only flush a buffer (i.e. clear the pending status) after it's done copying.

Reply:
But what about the `stream` argument?  We can leave `copyToMemoryBuffer` and unlock before the copy has actually completed, and if another copy occurs on a different stream (i.e.: DEFAULT_STREAM vs. the stream passed to `copyToMemoryBuffer`) I think we still could have issues.

Reply:
Added stream synchronize calls.